### PR TITLE
geolocation cleanup

### PR DESCRIPTION
**Pull request by schamp**
Surveyed all the geolocation code and found there was a lot of repetition.  Tried to refactor most of it to improve clarity, and reduce duplication.  The ultimate goal is to unlock on-demand loading of ZIP and adjacency data, and this cleanup moves us closer in that direction.

### CODE REVIEW

File: src/common/hooks/useGeolocatedRegions.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import useGeolocation from './useGeolocation';
+import useCountyToZipMap from './useCountyToZipMap';
+
+/**
+ * Hook to provide a map of region type (metroArea, county, state) to region
+ * based on the user's geolocation information.
+ * `isLoading` will be true until the data is available.
+ * `userRegions` will be null until loaded or if unavailable after loading
+ */
+const useGeolocatedRegions = () => {
```
Comment (schamp) [inline src/common/hooks/useGeolocatedRegions.ts:11]: This allows us to avoid calling `useGeolocation()` and `useCountyToZipMap()` in many places, which reduces the places we'll need to change to load just the necessary zip data.

File: src/common/hooks/useGeolocationInExplore.ts [hunk 2]
```diff
@@ -1,50 +0,0 @@
...
-  const [initialExploreFips, setInitialExploreFips] = useState(
-    getLargestMetroFipsForExplore(),
-  );
-
-  const { result: countyToZipMap } = useCountyToZipMap();
-
-  useEffect(() => {
-    if (geolocation && countyToZipMap) {
-      const userRegions = getGeolocatedRegions(geolocation, countyToZipMap);
-      if (!isNull(userRegions)) {
```
Comment (schamp) [inline src/common/hooks/useGeolocationInExplore.ts:36]: this filtering was moved into `filterGeolocatedRegions`

File: src/common/hooks/useGeolocationRegions.ts [hunk 3]
```diff
@@ -1,50 +0,0 @@
...
-  const isLoading = geoIsLoading || zipIsLoading;
-  // combine these and stringify so that useEffect will detect changes
-  const geolocationDataString = JSON.stringify(geolocationData);
-  const zipDataString = JSON.stringify(countyToZipMap);
-  useEffect(() => {
-    if (geolocationDataString && zipDataString) {
-      const geolocationInfo = JSON.parse(geolocationDataString);
-      const countyToZipMap = JSON.parse(zipDataString);
-
-      const userRegions = getGeolocatedRegions(geolocationInfo, countyToZipMap);
```
Comment (schamp) [inline src/common/hooks/useGeolocationRegions.ts:37]: this filtering was moved to `filterGeolocatedRegions`

File: src/common/regions/utils.ts [hunk 4]
```diff
@@ -166,17 +163,11 @@ export function getCountyRegionFromZipCode(
...
-  countyToZipMap: CountyToZipMap,
-): Region | undefined {
-  const countyFromZip = getCountyRegionFromZipCode(zipCode, countyToZipMap);
-  const metroFromZip =
-    countyFromZip &&
-    find(regions.metroAreas, (region: MetroArea) =>
-      region.countiesFips.includes((countyFromZip as County).fipsCode),
-    );
-  return metroFromZip;
+function getMetroRegionFromCounty(fipsCode: FipsCode): MetroArea | undefined {
```
Comment (schamp) [inline src/common/regions/utils.ts:166]: this is from an earlier change I was based on

File: src/common/regions/utils.ts [hunk 5]
```diff
@@ -206,17 +204,27 @@ export function filterUndefinedRegions(regions: any[]): Region[] | [] {
   );
 }
 
+// flattens and filters a map of region-type to regions to a list of valid regions
+export const filterGeolocatedRegions = (
```
Comment (schamp) [inline src/common/regions/utils.ts:208]: brings this duplicated code into once place

File: src/common/regions/utils.ts [hunk 6]
```diff
@@ -206,17 +204,27 @@ export function filterUndefinedRegions(regions: any[]): Region[] | [] {
...
+  }
+};
+
 /**
  * We rank a user's geolocated regions (state, county, metro if applicable) first
  * in the searchbar dropdown menu. This sorts the Region[] accordingly.
  */
 export function getAutocompleteRegionsWithGeolocation(
-  geolocation: GeolocationInfo,
+  geolocatedRegions: GeolocatedRegions,
```
Comment (schamp) [inline src/common/regions/utils.ts:225]: we can avoid passing the `countyToZipMap` by figuring out the `geolocatedRegions` at a higher level.

**Review by mikelehen — APPROVED** [review]
Sorry for the delay on this.  I don't know this code very well and it was a bit hard to follow how things were moving around (hopefully easier now with your changes 😁 ), plus I wanted to test it which took some extra time to figure out... but as far as I can tell this all makes sense and things still work! 🎉 

Sadly, you now have some merge conflicts to sort out.  Sorry. 😞  Hopefully that's straightforward but let me know if you want any extra eyes on it afterwards.

**Comment by schamp** [comment]
Yeah ha ha, it was pretty convoluted and took me a few tries to even figure out what made sense.  I'll rebase and fix the conflict, then merge.
